include/pybind11/docstring_options.h

```diff
@@ -0,0 +1,73 @@
+/*
+    pybind11/docstring_options.h: selectively suppress the automatic generation of docstrings and function signatures.
+
+    Copyright (c) 2016 Wenzel Jakob <wenzel.jakob@epfl.ch>
+
+    All rights reserved. Use of this source code is governed by a
+    BSD-style license that can be found in the LICENSE file.
+*/
+
+#pragma once
+
+#include "common.h"
+
+NAMESPACE_BEGIN(pybind11)
+
+class docstring_options
+{
+  public:
+
+    struct state {
+        bool show_user_defined; //< Include user-defined texts in docstrings.
+        bool show_signatures;   //< Inclcude auto-generated function signatures in docstrings.
+    };
+
+    PYBIND11_NOINLINE inline static state &global_state() {
+        static state instance = { true, true };
+        return instance;
+    }
+
+    docstring_options(bool show_all)
+    {
+        previous_state = global_state();
+        global_state().show_user_defined = show_all;
+        global_state().show_signatures = show_all;
+    }
+
+    docstring_options(bool show_user_defined, bool show_signatures)
+    {
+        previous_state = global_state();
+        global_state().show_user_defined = show_user_defined;
+        global_state().show_signatures = show_signatures;
+    }
+
+    ~docstring_options()
```

Comment:
> So what happens if it's heap-allocated and not destroyed at the end of the scope? Is it meant to never be created not on the stack?

It's basically just an RIAA-type class that overrides the globals while it exists then restores them when it gets destroyed.  So if it never gets destroyed, its override never gets restored (but this is basically true of any RAII class).  So I think the answer to the second question is "yes, usually", but I don't see that it would break if you managed it yourself on the heap (the worst case is that the values from before you instantiated it will never be restored).